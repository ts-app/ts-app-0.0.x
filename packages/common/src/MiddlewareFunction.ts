/**
 * Function called to process input and return output via the next() function when
 * Middleware.resolve() is called.
 *
 * @param {I} input passed to the middleware function
 * @param {(output?: O, options?: {resolved: boolean}) => void} next function is called to pass control to the next middleware
 * function, optionally passing in output data. To stop the middleware from further processing,
 * return  resolved=true
 * @param {O} prev output data from previous middleware or null if this is the first middleware
 * function
 */
export type MiddlewareFunction<I, O = any> = (input: I, next: (output?: O, options?: { resolved: boolean }) => void, prev?: O) => void
