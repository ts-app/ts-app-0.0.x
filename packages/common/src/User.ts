import { Profile } from './Profile'

export interface User<T extends Profile = Profile> {
  id: string
  createdAt: Date
  emails: {
    email: string
    verified: boolean
  }[]
  services?: {
    password?: {
      bcrypt: string
    }
    // resume?: {
    //   loginTokens: {
    //     when: Date
    //     token: string
    //   }[]
    // }
    // facebook?: {
    //   accessToken: string
    //   expiresAt: number
    //   id: string
    //   email: string
    //   name: string
    //   firstName: string
    //   lastName: string
    //   link: string
    // }
    jwt?: {
      refreshTokens: {
        name: string
        token: string
      }[]
    }
  }
  roles?: { role: string, group: string }[]
  profile?: T
}
