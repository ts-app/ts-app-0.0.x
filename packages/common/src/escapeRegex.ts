/**
 * Escape user input to be treated as a literal string within a regular expression.
 *
 * Source: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
 *
 * @param {string} s
 */
export const escapeRegex = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
