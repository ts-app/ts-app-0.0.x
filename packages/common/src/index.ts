export * from './escapeRegex'
export * from './loadFile'
export * from './Middleware'
export * from './MiddlewareFunction'
export * from './Profile'
export * from './User'
