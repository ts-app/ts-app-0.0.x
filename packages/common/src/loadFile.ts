import * as fs from 'fs'
import * as path from 'path'

/**
 * Utility function to load contents of specified file and convert to UTF-8 string.
 *
 * @param {string} filename filename to load
 * @return {string} UTF-8 string
 */
export const loadFile = (filename: string) => {
  const filePath = path.resolve(filename)
  return fs.readFileSync(filePath, 'utf8').toString()

// tslint:disable-next-line
}
