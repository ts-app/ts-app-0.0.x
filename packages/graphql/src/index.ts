export * from './demoSchemaDefinition'
export * from './DemoService'
export * from './environmentVariableIsSet'
export * from './mergeSchemaDefinitions'
export * from './mergeTypeDefs'
export * from './Resolver'
export * from './ResolverError'
export * from './ResolverMiddlewareInput'
export * from './ResolverService'
export * from './SchemaDefinition'
export * from './SecurityMiddleware'
export * from './serveSchema'
export * from './ServiceInfo'
export * from './standardSchemaDefinition'
export * from './userIdMatchParam'
