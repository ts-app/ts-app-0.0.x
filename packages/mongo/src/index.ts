export * from './FindWithCursor'
export * from './MongoService'
export * from './MongoUpdateBuilder'
