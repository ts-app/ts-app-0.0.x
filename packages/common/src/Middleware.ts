import { MiddlewareFunction } from './MiddlewareFunction'

/**
 * A middleware/filter pattern. Current implementation works like a filter where every registered
 * middleware function will be executed when resolve() is called.
 *
 * Filter: https://en.wikipedia.org/wiki/Intercepting_filter_pattern
 * Middleware: https://dzone.com/articles/understanding-middleware-pattern-in-expressjs
 *
 */
export class Middleware<I, O = any> {
  constructor (private middlewares: MiddlewareFunction<I, O>[] = []) {
  }

  use (...middlewares: MiddlewareFunction<I, O>[]) {
    this.middlewares.push(...middlewares)
  }

  resolve (param: I) {
    const middlewares = [ ...this.middlewares ].reverse()
    const current = middlewares.pop()
    return this.next(param, middlewares, current)
  }

  private async next (param: I, middlewares: MiddlewareFunction<I, O>[], currentMiddlewareFn?: MiddlewareFunction<I, O>, prev?: O): Promise<any> {
    if (currentMiddlewareFn) {
      return new Promise((resolve, reject) => {
        const next = (prevOutput?: O, options?: { resolved: boolean }) => {
          if (options && options.resolved === true) {
            // stop calling subsequent middleware function
            resolve(prevOutput)
          } else {
            // call next middleware function
            const middlewareFn = middlewares.pop()
            this.next(param, middlewares, middlewareFn, prevOutput).then(resolve).catch(reject)
          }
        }
        const fn: any = currentMiddlewareFn(param, next, prev)

        // if middleware function is a promise, catch exception and reject from outer promise
        if (typeof fn === 'object' && typeof fn.then === 'function') {
          fn.catch(reject)
        }
      })
    } else {
      return prev
    }
  }
}
