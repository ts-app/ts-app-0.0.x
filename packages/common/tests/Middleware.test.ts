import { Middleware } from '../src/Middleware'
import { MiddlewareFunction } from '../src/MiddlewareFunction'

type Input = { amount: number }
type Output = string

describe('Middleware', () => {
  test('resolve() without middleware function', async () => {
    expect(await new Middleware().resolve({})).toBeUndefined()
  })

  test('resolve() single middleware function', async () => {
    const m = new Middleware<Input>([
      (input, next) => next(input.amount * 123)
    ])

    expect(await m.resolve({ amount: 50 })).toBe(6150)
  })

  test('resolve() two middleware functions, get prev value, set next value', async () => {
    const addAmount: MiddlewareFunction<Input, Output> = (param, next) => {
      param.amount += 5
      next('from to add')
    }
    const asyncMultipleAmount: MiddlewareFunction<Input, Output> = (param, next, prev) => {
      // got prev value from middlewareToAdd
      expect(prev).toBe('from to add')

      setTimeout(() => {
        param.amount *= 3
        next(`Amount is ${param.amount}`)
      }, 0)
    }

    const middleware = new Middleware<Input>([
      addAmount,
      asyncMultipleAmount
    ])

    const param = { amount: 10 }
    const output = await middleware.resolve(param)
    expect(param.amount).toBe(45)
    expect(output).toBe('Amount is 45')
  })

  test('prev value with multiple middleware functions', async () => {
    const m = new Middleware<Input>([
      (input, next, prev) => {
        expect(prev).toBeUndefined()
        next(123)
      },
      (input, next, prev) => {
        expect(prev).toBe(123)
        next(456)
      },
      (input, next, prev) => {
        expect(prev).toBe(456)
        next(789)
      }
    ])

    const resolved = await m.resolve({ amount: 10 })
    expect(resolved).toBe(789)
  })

  test('async middleware function', async () => {
    const m = new Middleware<Input>([
      (input, next, prev) => {
        expect(prev).toBeUndefined()
        next(123)
      },
      async (input, next, prev) => {
        expect(prev).toBe(123)
        setTimeout(() => next(456), 0)
      },
      (input, next, prev) => {
        expect(prev).toBe(456)
        next(789)
      }
    ])

    const resolved = await m.resolve({ amount: 10 })
    expect(resolved).toBe(789)
  })

  test('middleware throws error', async () => {
    // once a middleware throws error, it must be handled by resolve()
    // should we allow subsequent middleware to be called? maybe, but it gets complicated. let's not.
    const middleware = new Middleware<Input>([
      (input, next) => next(123),
      () => {
        throw new Error('Simulated error')
      },
      () => fail('Should not reach this line')
    ])

    try {
      await middleware.resolve({ amount: 10 })
      fail('Should not reach this line')
    } catch (e) {
      // handle error
      expect(e).toEqual(new Error('Simulated error'))
    }
  })

  test('async middleware throws error', async () => {
    // once a middleware throws error, it must be handled by resolve()
    // should we allow subsequent middleware to be called? maybe, but it gets complicated. let's not.
    const middleware = new Middleware<Input>([
      (input, next) => next(123),
      async () => {
        throw new Error('Simulated error')
      },
      () => fail('Should not reach this line')
    ])

    try {
      await middleware.resolve({ amount: 10 })
      fail('Should not reach this line')
    } catch (e) {
      // handle error
      expect(e).toEqual(new Error('Simulated error'))
    }
  })

  test('middleware function can prevent calling subsequent functions', async () => {
    const flow: string[] = []

    const m = new Middleware<Input>([
      (input, next, prev) => {
        flow.push('first')
        next(123)
      },
      (input, next, prev) => {
        flow.push('second')
        next(456, { resolved: true })
      },
      (input, next, prev) => {
        flow.push('third')
        next(789)
      }
    ])

    const resolved = await m.resolve({ amount: 10 })
    expect(resolved).toBe(456)
    expect(flow).toMatchSnapshot()
  })
})
