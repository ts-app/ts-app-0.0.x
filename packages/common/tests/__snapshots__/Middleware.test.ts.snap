// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware middleware function can prevent calling subsequent functions 1`] = `
Array [
  "first",
  "second",
]
`;
